### ISSUE TITLE
[yulopti] interactively run one iteration of the sequence between [] from fullsuite

### ISSUE THREAD

**Issue by hrkrshnn** [ethereum/solidity#10812]
If fullsuite is

```
		"dhfoDgvulfnTUtnIf"            // None of these can make stack problems worse
		"["
			"xarrscLM"                 // Turn into SSA and simplify
			"cCTUtTOntnfDIul"          // Perform structural simplification
			"Lcul"                     // Simplify again
			"Vcul jj"                  // Reverse SSA

			// should have good "compilability" property here.

			"Tpeul"                    // Run functional expression inliner
			"xarulrul"                 // Prune a bit more in SSA
			"xarrcL"                   // Turn into SSA again and simplify
			"gvif"                     // Run full inliner
			"CTUcarrLsTFOtfDncarrIulc" // SSA plus simplify
		"]"
		"jmuljuljul VcTOcul jmul";     // Make source short and pretty
```

I'm referring to the part
```
	"["
			"xarrscLM"                 // Turn into SSA and simplify
			"cCTUtTOntnfDIul"          // Perform structural simplification
			"Lcul"                     // Simplify again
	....
        ....
        "]"
```
------

Based on https://github.com/ethereum/solidity/issues/10812#issuecomment-852787213.

1. Assign a key inside `yulopti`: instead of running a single optimizer step, run the full inner loop (i.e., the sequence inside `[]`) of the default optimizer sequence.
2. `yulopti` should check the sequence find where the loop is and run all the steps in this loop in the order they are given.
3. The output would be printed after one round of the loop instead of each step.

(As a bonus, it would also be nice to have keys that correspond to the prefix and postfix part of the loop.)

**Comment by hrkrshnn** [comment]
@cameel What do you think?

**Comment by hrkrshnn** [comment]
For some context: https://github.com/ethereum/solidity/pull/10418#discussion_r559664921

**Comment by chriseth** [comment]
Does it need to be interactive? The optimizer already has the option to output the code whenever there was a change. We could change this to output the code not after each step that did a change, but only after one loop iteration. Would that help?

**Comment by cameel** [comment]
Or we could add an option like `--yul-optimizations` to yulopti to make it run the sequence you specify print optimized code and exit.

**Comment by cameel** [comment]
Actually, you can already do that using `yul-phaser`. I implemented a feature to print optimized programs for debugging and you can tell it not to mutate anything and exit after 0 rounds:

```bash
yul-phaser \
    --mode print-optimised-programs \
    --rounds 0 \
    --show-initial-population \
    --population "" \
    --population xarrscLM \
    test/libyul/yulOptimizerTests/fullInliner/simple.yul
```

Output:
```yul
Chromosome:
Program: test/libyul/yulOptimizerTests/fullInliner/simple.yul
{
    {
        let y := add(f(sload(mload(2))), mload(7))
    }
    function f(a) -> x
    {
        let r := mul(a, a)
        x := add(r, r)
    }
}
Chromosome: xarrscLM
Program: test/libyul/yulOptimizerTests/fullInliner/simple.yul
{
    {
        let _1 := 7
        let _2 := mload(_1)
        let _3 := 2
        let _4 := mload(_3)
        let _5 := sload(_4)
        let _6 := f(_5)
        let y := add(_6, _2)
    }
    function f(a) -> x
    {
        let r := mul(a, a)
        let x_7 := add(r, r)
        x := x_7
    }
}
```

The `--population` option accepts an optimization sequence. It's meant to be used multiple times to specify the whole initial population but in the example above I'm giving it `""` to see the unoptimized program and `xarrscLM` as an example sequence to check out. You can also put the sequence(s) in a file (`--population-from-file`), one sequence per line. The be precise, unoptimized program is not not completely unoptimized - phaser applies diambiguator, function hoister, function grouper and for loop init rewriter to every program so you may want to skip these from your sequence.

The only downside is that phaser does not support square bracket notation (might support that in the future if we really want it but mutating that is tricky) so you just have to repeat parts of the sequence manually.

**Comment by chriseth** [comment]
Can you clarify a bit what this issue is about? Assigning a key inside `yulopti` to instead of running a single optimizer step, run the full inner loop of the default optimizer sequence? If yes, this would mean you have to check this sequence, find where the loop is and then run all the steps in this loop in the order they are given, right? And the idea is that the modified code is only output once at the end and not after each step, right?

**Comment by hrkrshnn** [comment]
The original goal was to see the optimized output after one round of the sequence inside square bracket.

> Assigning a key inside yulopti to instead of running a single optimizer step, run the full inner loop of the default optimizer sequence
> If yes, this would mean you have to check this sequence, find where the loop is and then run all the steps in this loop in the order they are given, right? And the idea is that the modified code is only output once at the end and not after each step, right?

This sounds good.

I'll update the issue description with this.
